javascript/ql/src/Performance/ReDoS.ql

```diff
@@ -705,7 +705,25 @@ predicate isFork(State q, InputSymbol s1, InputSymbol s2, State r1, State r2) {
     or
     r1 = r2 and
     q1 = q2 and
-    epsilonSucc+(q) = q
+    epsilonSucc+(q) = q and
+    exists(RegExpTerm term | term = q.getRepr() | term instanceof InfiniteRepetitionQuantifier) and
```

Comment:
I'm not seeing the rationale behind this whole epsilon-loop case.

RegExp engines will generally not allow the body of a repetition clause to match the empty string. Attempting to match `E*` while allowing `E` to match the empty string immediately leads to an infinite loop.

A fork that matches a loop of epsilon edges therefore looks suspicious to me, as this loop is likely a spurious artifact of our own NFA construction.

I've been trying to think of what cases we could be missing here that would explain the FNs you saw. Here's an idea:

Could it be that there are two distinct epsilon-paths (not necessarily looping) leading from `q` to `q1=q2`, and if the engine failed after trying the first path, it will retry using the other path? (which is futile of course, as it's in the same state, but the engine doesn't check for this).

(When I say _distinct_ paths, it's possible they actually need to be _disjoint_ paths.)

Reply:
It's to flag something like `/(a?b?)*c/` ([debug session](https://regex101.com/r/30l4PM/2/debugger)).   
Where the body inside the repetition does match the empty-string.    
This causes the regexp engine to try to match against every possible subset of the input string (see the debug session).  

I'll look more into it, but I don't see another way to flag it.

Reply:
I think the general pattern is that the body of the repetition can match something of different lengths, where there is no way for the engine to determine where one repetition starts/ends compared to the next repetition.    

In the above example, the body matches something of length between 0 and 2. 

Consider this example (which we don't flag): `var bad = /(a(\w)?)*X/`.   
The inner body either matches `a`, or `aa`.  
But these combinations of strings of length 1 or 2 can be combined in `O(2^n)` possible ways, giving us exponential backtracking.     

At least one of the regexp [just fixed in highlight.js](https://github.com/highlightjs/highlight.js/commit/373b9d862401162e832ce77305e49b859e110f9c#diff-5819541c636e70dcfc768352222f083986ce06ab582b657dca8cb50364e2a334L34) was of this pattern.

Reply:
I'll revert the epsilon idea for this PR.   

I'll look into it for (yet another) followup PR.

Reply:
Ok, I think I see how it works for `/(a?b?)*c/`. The fork state is right after reading `a`, as there are two different ways to read `b`: one by skipping `b?` and looping around, and the other by immediately consuming `b?`, but we missed that fork because both lead to the same NFA state. I can see the relevance of epsilon loop now. I'm still not sure if it has to be a loop or just two distinct paths, but matching loops as a safe approximation is fine with me.

I'm still not sure about the second commit that added the extra conditions.

I'll leave it up to you if you'd prefer to use this PR or in another one.

Reply:
Actually maybe we're on to something here. The FPs you observed from the _first_ version of this check could be from the following spurious fork: Before reading `a?` there are seemingly two ways to read `a`: skipping `a?`and looping around; or immediately consuming `a?`. Exactly as with the `b` fork, except this one is spurious because it involves a repetition body matching an empty string, which won't ever happen.

Reply:
> I'll leave it up to you if you'd prefer to use this PR or in another one.

I'll wait until another PR.   
I also found another thing I want to put in the next PR. 

> Ok, I think I see how it works for /(a?b?)*c/.   

I'm still not sure.   
When I try it out, the attack string must be repetitions of `ab`, otherwise the attack doesn't work.   
Just repeating `a` or `b` doesn't work.   

> I'm still not sure about the second commit that added the extra conditions.

Me too. But it was the best I could come up with.